java/java-psi-api/src/com/intellij/psi/util/TypeConversionUtil.java

```diff
@@ -853,6 +853,12 @@ private static boolean isAssignable(@NotNull PsiType left,
              && rText.endsWith(lText)
              && rText.charAt(rText.length() - lText.length() - 1) == '.';
     }
+
+    final PsiClass leftClass = PsiUtil.resolveClassInType(left);
+    if (leftClass instanceof PsiTypeParameter) {
```

Comment:
Currently I assume that I can assign any type to a type parameter. This solution allows `Intellij IDEA` add to the list of inferred types `Function`, `BiFunction`, (any `@FunctionalInterface` with generic return type)

This idea might not be appropriate, so I require guidance from the `jetbrains` team.

Reply:
I am afraid, the approach is wrong: at very least type parameters can have bounds and such change would mean lots of bad code green/good code red problems. I suggest not to change this utility, cause changes here would influence the whole java type system

Reply:
Thank you for your response, I do realize that this change is wrong, but it covers the case with generic types and I just need to figure out how to handle this case. Furthermore after this change IDEA starts suggesting more `@FunctionalInterface` types when a local variable is extracted (I described it in the issue). I assume there has to be a method that handles unification of generic types, it would be great if the `JetBrains` team guides me.
I would really like to complete the PR, because this feature is needed in my `Intellij IDEA` plugin.